src/DynamoDbConnection.php

```
@@ -0,0 +1,282 @@
+            $sortKey = null;
+        }
+
+        if (!$partitionKey) {
+            throw new InvalidArgumentException('A partition key is required.');
+        }
+
+        if ($sortKey && !$this->sortKeyAttribute) {
+            throw new InvalidArgumentException('A sort key attribute must be defined to use compound keys.');
+        }
+
+        $value = [
+            $this->partitionKeyAttribute => $partitionKey,
+        ];
+
+        if ($this->sortKeyAttribute) {
+            if ($sortKey === null) {
+                throw new InvalidArgumentException('A sort key attribute was specified, but no value was found.');
+            }
+            $value[$this->sortKeyAttribute] = $sortKey;
+        }
+
+        return $this->marshaler->marshalItem($value);
+    }
+
+    protected function getClient(): DynamoDbClient
+    {
+        if ($this->client) {
+            return $this->client;
+        }
+
+        $config = array_filter([
+            'credentials' => $this->getCredentials(),
+            'region' => $this->getRegion(),
+            'version' => $this->version,
+            'endpoint' => $this->endpoint,
+        ]);
+
+        $this->client = new DynamoDbClient($config);
+
+        return $this->client;
+    }
+
+    protected function getCredentials(): Credentials|callable
+    {
+        return $this->_credentials ?? CredentialProvider::defaultProvider();
+    }
+
+    protected function setCredentials($credentials): void
+    {
+        if (is_array($credentials)) {
+            $credentials = new Credentials(
+                $credentials['key'],
+                $credentials['secret'],
+                $credentials['token'] ?? null,
+                $credentials['expires'] ?? null,
+            );
+        }
+
+        $this->_credentials = $credentials;
+    }
+
+    protected function setRegion(string $region): void
+    {
+        $this->_region = $region;
+    }
+
+    protected function getRegion(): ?string
```

Probably not needed, but if none of these are set, should we throw an exception?

---

We could , but region is required when creating the client, so it will already throw a nice exception for us.